gobblin-service/src/main/java/org/apache/gobblin/service/modules/flowgraph/FlowGraph.java

```diff
@@ -0,0 +1,102 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.flowgraph;
+
+import java.util.Collection;
+
+import org.apache.gobblin.annotation.Alpha;
+
+
+/**
+ * An interface for {@link FlowGraph}. A {@link FlowGraph} consists of {@link DataNode}s and {@link FlowEdge}s.
+ * The interface provides methods for adding and removing {@link DataNode}s and {@link FlowEdge}s to the {@link FlowGraph}.
+ * In addition the interface provides methods to return factory classes for creation of {@link DataNode}s and {@link FlowEdge}s.
+ */
+
+@Alpha
+public interface FlowGraph {
+  public Collection<DataNode> getNodes();
+
+  /**
+   * Add a {@link DataNode} to the {@link FlowGraph}
+   * @param node {@link DataNode} to be added
+   * @return true if {@link DataNode} is added to the {@link FlowGraph} successfully.
+   */
+  public boolean addDataNode(DataNode node);
+
+  /**
+   * Add a {@link FlowEdge} to the {@link FlowGraph}
+   * @param edge {@link FlowEdge} to be added
+   * @return true if {@link FlowEdge} is added to the {@link FlowGraph} successfully.
+   */
+  public boolean addFlowEdge(FlowEdge edge);
+
+  /**
+   * Remove a {@link DataNode} and all its incident edges from the {@link FlowGraph}
+   * @param node {@link DataNode} to be removed
+   * @return true if {@link DataNode} is removed from the {@link FlowGraph} successfully.
+   */
+  public boolean deleteDataNode(DataNode node);
+
+  /**
+   * Remove a {@link DataNode} and all its incident edges from the {@link FlowGraph}
+   * @param nodeId identifier of the {@link DataNode} to be removed
+   * @return true if {@link DataNode} is removed from the {@link FlowGraph} successfully.
+   */
+  public boolean deleteDataNode(String nodeId);
+
+  /**
+   * Remove a {@link FlowEdge} from the {@link FlowGraph}
+   * @param edge to be removed
+   * @return true if edge is removed from the {@link FlowGraph} successfully.
+   */
+  public boolean deleteFlowEdge(FlowEdge edge);
+
+  /**
+   * Remove a {@link FlowEdge} from the {@link FlowGraph}
+   * @param edgeId label of the edge to be removed
+   * @return true if edge is removed from the {@link FlowGraph} successfully.
+   */
+  public boolean deleteFlowEdge(String edgeId);
+
+  /**
+   * Get the factory class for creating {@link DataNode}s to be added to the {@link FlowGraph}.
+   * @return a {@link DataNodeFactory} implementation.
+   */
+  public DataNodeFactory getDataNodeFactory();
```

Comment:
`DataNode`s are quite heterogeneous in a graph. Different `DataNode` will be created in different ways. We should consider removing this behavior.

Managing the creation of different DataNodes beyonds the responsiblily of a graph, which only manages the in-and-out of a node and discovers path

Reply:
DataNodes and FlowEdges are homogeneous. Each DataNode/FlowEdge in the graph is created in a common way. DataNodes and FlowEdges are distinguishable by their properties. 

The idea of providing node/edge creation factories is to hide the details of the creation process to the external classes.